pkg/cmd/bootstrap.go

```diff
@@ -168,34 +176,82 @@ func initiateInteractiveMode(io *BootstrapParameters, client *utility.Client) er
 		identifier := factory.NewDriverIdentifier(factory.Mapping(host, io.PrivateRepoDriver))
 		factory.DefaultIdentifier = identifier
 	}
-	if ui.UseInternalRegistry() {
-		io.ImageRepo = ui.EnterImageRepoInternalRegistry()
-	} else {
-		io.ImageRepo = ui.EnterImageRepoExternalRepository()
-		io.DockerConfigJSONFilename = ui.EnterDockercfg()
+	if cmd.Flag(imageRepoFlag).Changed {
+		isInternalRegistry, _, err := imagerepo.ValidateImageRepo(io.ImageRepo)
+		if !isInternalRegistry {
+			if !cmd.Flag("dockercfgjson").Changed {
```

Comment:
This doesn't quite look right.

`dockercfgjson` has a default value, and https://godoc.org/github.com/spf13/pflag#FlagSet.Changed returns true if it's changed.

If the user doesn't change the default, they'll get prompted to enter it, even if they wanted to keep the default.

My gut feeling, is that this should be `if io.DockerConfigJSONFilename == "" {` i.e. if it's an external repo, but the auth has been cleared, prompt for a Docker config file?

Reply:
Good catch. I'm missing the `&& promptForAll` here.  It should be:

`if !cmd.Flag("dockercfgjson").Changed && promptForAll {`

For not accepting defaults, you will only get prompted.  This part of the code is where we try to be smart and prompt for the config file when necessary.

In this example, the external image repo is provided on the command line, defaults are NOT accepted, and the config file is NOT overridden, so you will get prompted.
```
kam bootstrap --service-repo-url https://github.com/keithchong/taxi.git --image-repo quay.io/keithchong/taxi --interactive
  
Checking dependencies

 ✓  Checking if Sealed Secrets is installed with the default configuration [189ms]
 ✓  Checking if ArgoCD Operator is installed with the default configuration [437ms]
 ✓  Checking if OpenShift Pipelines Operator is installed with the default configuration [511ms]
  
Starting interactive prompt

? Do you want to accept all default values and be prompted only for the minimum required options? no
? Provide the URL for your GitOps repository https://github.com/keithchong/gitops.git
  The supplied image repository has been detected as an external repository.
? Provide the path to config.json which authenticates image pushes to the desired image registry [? for help] (~/.docker/config.json) 
```

And with the change `&& promptForAll` added, then when you do accept defaults, the third question above will NOT be asked.  That's the only difference.


For the last paragraph of your comment, that means the user specified the flag from the command line to be empty....or it could be an invalid location or it could be pointing to some non-config file.  I don't think it matters at this point of the interactive session. Later on, the authentication will simply fail.  So you will have to 'start over'.

So the gist of this part of the change is this.  If you specified the `--image-repo` flag from the command line, and if you are in interactive mode, we will determine if the image repo is external.  If external, we will prompt you for the location of the config file.  This prompting depends on if you have specified that flag from the command line or if you decided to accept defaults or not.  And the value of the config file location will be validated later.  It's a 'don't care' at this point, since the validation is done during `Run`